### PR TITLE
tools for analyzing, updating and adding copyright headers in source files

### PR DESCRIPTION
**Pull request by isle2983**
Three scripts in contrib/devtools/ are introduced:

**copyright_header_report.py**
- outputs a report of copyright claims in the source files
- helps to identify files that are missing copyright
- helps to identify files that were added that have a new copyright holder beyond the current known set
- helps to find typos in the copyright header

**copyright_header_update.py**
- replaces fix-copyright-headers.py
- does file editing in native python rather than an external call to perl

**copyright_header_insert.py**
- inserts a properly-formatted header with proper dates into a given source file

Altogether, I hope they make the task of managing the copyright headers a bit
easier.

### CODE REVIEW

**Comment by fanquake** [comment]
I'm not sure that maintaining copyright headers is such an issue that it warrants splitting one tool into three. Interested in other opinions.

**Comment by MarcoFalke** [comment]
Agree with @fanquake: I'd prefer to have the functionality (report and update) in a single script.

 Also, I don't think **copyright_header_insert.py** is required. (If someone forgets to add the header, they will forget to run the script as well)

**Comment by isle2983** [comment]
I have condensed the three files into one and made them subcommands:
```
$ ./copyright_header_report.py <base_directory> [verbose]
$ ./copyright_header_update.py <base_directory>
$ ./copyright_header_report.py <file>
```

becomes:

```
$ ./copyright_header.py report <base_directory> [verbose]
$ ./copyright_header.py update <base_directory>
$ ./copyright_header.py insert <file>
```

running with no subcommand to copyright_header.py lists a usage string with the subcommands. Running a subcommand without arguments displays the subcommands usage string.


If we are against the 'insert' functionality, I will remove it, however I think
it is useful for a few things not mentioned:

1) It defines the exact format of a proper header
2) It gets the date right
3) It can be a building block of a strict CI system that rejects bad/missing headers and directs you to "please run` ./copyright_header.py insert <your file>`" to fix the problem.

**Comment by fanquake** [comment]
You'll also need to update the contrib docs [here](https://github.com/bitcoin/bitcoin/blob/master/contrib/devtools/README.md#fix-copyright-headerspy).

**Comment by isle2983** [comment]
Thanks @fanquake. 159597a2b8967c8bc8106d98705d68932500b58a includes an update to README.md to document the new script and subcommands. Also, I tweaked some grammar and fixed some spelling in the usage strings inside copyright_header.py for clarity and made sure they matched README.md in content.

[Rendered README.md](https://github.com/isle2983/bitcoin/blob/copyright-scripts/contrib/devtools/README.md#copyright_headerpy)

**Comment by laanwj** [comment]
Concept ACK 159597a

File: contrib/devtools/copyright_header.py [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+    'src/tinyformat.h',
+    'src/leveldb/util/env_win.cc',
+    'src/crypto/ctaes/bench.c',
+    'qa/rpc-tests/test_framework/bignum.py',
+    # python init:
+    '*__init__.py',
+]
+EXCLUDE_COMPILED = re.compile('|'.join([fnmatch.translate(m) for m in EXCLUDE]))
+
+INCLUDE = ['*.h', '*.cpp', '*.cc', '*.c', '*.py']
```
Comment (luke-jr) [inline contrib/devtools/copyright_header.py]: Probably should include *.sh, *.am, *.m4, and *.include

Might make more sense as an exclusion list...

Comment (MarcoFalke) [inline contrib/devtools/copyright_header.py]: We don't use `cc`

Comment (laanwj) [inline contrib/devtools/copyright_header.py]: Leveldb does, though, but yes it makes no sense to include that here.

**Comment by rebroad** [comment]
Who should run this and when?

**Comment by isle2983** [comment]
Good question.

I have a working branch that hooks it into TravisCI to fail branches that are missing copyright headers - forcing the submitter to either add them or to manually exclude the file from the check. I plan on submitting the PR when I get the chance to polish it a bit more

If the overall feeling is that this kind of thing is helpful and not a pain for day-to-day work, there might be utility in doing similar scripts for TravisCI to gradually ratchet in enforcement of basic whitespace rules and/or eventually clang-format and/or pylint style checks.

Also, the years in the headers need to be upgraded every so often to make sure they are current. December 31st might be a good day to run the script and post the PR.

**Comment by laanwj** [comment]
> Who should run this and when?

- The best time to update a copyright header is when you touch the file
- The second-best is a single mass update at the end of the year of all the files that were touched in that year (that's what the scripting is for).

**Pull request merged**